crates/arroyo-controller/src/states/rescaling.rs

```diff
@@ -12,59 +16,27 @@ impl State for Rescaling {
     }
 
     async fn next(mut self: Box<Self>, ctx: &mut JobContext) -> Result<Transition, StateError> {
-        let job_controller = ctx.job_controller.as_mut().unwrap();
-
-        let mut final_checkpoint_started = false;
-
-        loop {
-            match job_controller.checkpoint_finished().await {
-                Ok(done) => {
-                    if done && job_controller.finished() && final_checkpoint_started {
-                        return Ok(Transition::next(*self, Scheduling {}));
-                    }
-                }
-                Err(e) => {
-                    return Err(ctx.retryable(
-                        self,
-                        "failed while monitoring final checkpoint",
-                        e,
-                        10,
-                    ));
-                }
-            }
-
-            if !final_checkpoint_started {
-                match job_controller.checkpoint(true).await {
-                    Ok(started) => final_checkpoint_started = started,
-                    Err(e) => {
-                        return Err(ctx.retryable(
-                            self,
-                            "failed to initiate final checkpoint",
-                            e,
-                            10,
-                        ));
-                    }
-                }
-            }
-
-            match ctx.rx.recv().await.expect("channel closed while receiving") {
-                JobMessage::RunningMessage(msg) => {
-                    if let Err(e) = job_controller.handle_message(msg).await {
-                        return Err(ctx.retryable(
-                            self,
-                            "failed while waiting for job finish",
-                            e,
-                            10,
-                        ));
-                    }
-                }
-                JobMessage::ConfigUpdate(c) => {
-                    stop_if_desired_non_running!(self, &c);
-                }
-                _ => {
-                    // ignore other messages
-                }
+        match ctx.take_stopping_checkpoint().await {
```

Comment:
It looks like there might be some problems with rescaling. I got into a state where the the operators didn't shutdown cleanly, getting the controller stuck indefinitely in rescaling:
[rescaling.txt](https://github.com/ArroyoSystems/arroyo/files/14728293/rescaling.txt).

I also wasn't able to force stop out of rescaling — the pipeline was just wedged until I restarted the controller, then on recovery it went into CheckpointStopping where it also got stuck.

Reply:
I'm able to consistently reproduce this by doing two rescales in a row